include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -168,39 +317,115 @@ class __sort_global_kernel;
+        // __n is taken as is because of the bit floor and processing 2 items per work-item
+        // hence the processed size always fits a single work-group if __n is chosen
+        __wg_size = std::min<std::size_t>({__wg_size, __slm_max_wg_size, __n});
+        __wg_size = oneapi::dpl::__internal::__dpl_bit_floor(__wg_size);
+        _Leaf2 __leaf_sorter(__rng, __comp, __wg_size);
+        return __caller(std::forward<_ExecutionPolicy>(__exec), __rng, __comp, __leaf_sorter);
+    }
+};
+
 template <typename _ExecutionPolicy, typename _Range, typename _Compare>
 auto
 __parallel_sort_impl(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec, _Range&& __rng,
                      _Compare __comp)
 {
-    using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
-
-    const auto __n = __rng.size();
-    if (__n <= std::numeric_limits<std::uint32_t>::max())
+    if (__rng.size() <= std::numeric_limits<std::uint32_t>::max())
     {
-        using _wi_index_type = std::uint32_t;
-        using _LeafSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_leaf_kernel<_CustomName, _wi_index_type>>;
-        using _GlobalSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_global_kernel<_CustomName, _wi_index_type>>;
-        using _CopyBackKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_copy_back_kernel<_CustomName, _wi_index_type>>;
-        return __parallel_sort_submitter<_wi_index_type, _LeafSortKernel, _GlobalSortKernel, _CopyBackKernel>()(
-            oneapi::dpl::__internal::__device_backend_tag{}, std::forward<_ExecutionPolicy>(__exec),
-            std::forward<_Range>(__rng), __comp);
+        return __submit_selecting_leaf<std::uint32_t>(std::forward<_ExecutionPolicy>(__exec),
+                                                      std::forward<_Range>(__rng), __comp);
     }
     else
     {
-        using _wi_index_type = std::uint64_t;
-        using _LeafSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_leaf_kernel<_CustomName, _wi_index_type>>;
-        using _GlobalSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_global_kernel<_CustomName, _wi_index_type>>;
-        using _CopyBackKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
-            __sort_copy_back_kernel<_CustomName, _wi_index_type>>;
-        return __parallel_sort_submitter<_wi_index_type, _LeafSortKernel, _GlobalSortKernel, _CopyBackKernel>()(
-            oneapi::dpl::__internal::__device_backend_tag{}, std::forward<_ExecutionPolicy>(__exec),
-            std::forward<_Range>(__rng), __comp);
+        return __submit_selecting_leaf<std::size_t>(std::forward<_ExecutionPolicy>(__exec), std::forward<_Range>(__rng),
```

Comment:
May be better to use `uint32_t` type?

Reply:
Did you mean `std::uint64_t` since `std::size_t` may be even "smaller" than `std::uint32_t` above?

Reply:
Yes, exactly, i believe then `std::uint64_t` for cases when `std::uint32_t` is not enough.

Reply:
Done